### PR TITLE
Add align attribute for params pointers

### LINKED ISSUE #671: add align <n> attribute to pointer parameters which have bigger than normal alignment
http://llvm.org/docs/LangRef.html#parameter-attributes
> align <n>
  >   This indicates that the pointer value may be assumed by the optimizer to have the specified alignment.
   >  Note that this attribute has additional semantics when combined with the byval attribute.

```zig
export fn entry() {
    var a: u8 align(16) = 2;
    bar(&a);
}
fn bar(ptr: &align(16) u8) {

}
pub fn panic(msg: []const u8) -> noreturn { unreachable; }
```

Generates this LLVM:

```llvm
define internal fastcc void @bar(i8* nonnull) unnamed_addr #2 !dbg !34 {
```

but it should be:

```llvm
define internal fastcc void @bar(i8* nonnull align<16>) unnamed_addr #2 !dbg !34 {
```

This is a missed optimization opportunity.

### PR DESCRIPTION
**Pull request by LemonBoy**
Fixes #671

### CODE REVIEW

File: src/codegen.cpp [hunk 1]
```diff
@@ -2279,6 +2279,9 @@ void walk_function_params(CodeGen *g, ZigType *fn_type, FnWalk *fn_walk) {
                 if ((param_type->id == ZigTypeIdPointer && param_type->data.pointer.is_const) || is_byval) {
                     addLLVMArgAttr(llvm_fn, (unsigned)gen_index, "readonly");
                 }
+                if (param_type->id == ZigTypeIdPointer && param_type->data.pointer.explicit_alignment != 0) {
```
Comment (andrewrk) [inline src/codegen.cpp]: We can do even better than `param_type->id == ZigTypeIdPointer`, handling `?*align(n) T` (optional pointers) as well. I believe you want `get_codegen_ptr_type(param_type) != nullptr`. The function you already chose, `get_ptr_align`, supports any type for which `get_codegen_ptr_type` returns non-null.

When you do this, accessing `data.pointer.explicit_alignment` however will only work for `ZigTypeIdPointer`, so there are a couple options:
 * Put the `align <n>` attribute on all pointer parameters. I'm sure this is fine.
 * Introduce a new function in analyze.cpp, `get_ptr_align_if_not_abi_aligned` which returns the alignment, or 0 if ABI aligned, in which case you can avoid adding the attribute.

Comment (LemonBoy) [reply src/codegen.cpp]: > Put the align <n> attribute on all pointer parameters. I'm sure this is fine.

I went for the easiest solution, after all llvm is going to compute the alignment anyway.

**Comment by LemonBoy** [comment]
Correct me if I'm wrong but I don't see any testing harness to check the emitted LLVM IR, hence no test case for this :)

**Comment by andrewrk** [comment]
That's correct. Maybe someday we'll have tests for that, but for now this is OK :+1:

**Pull request merged**